### PR TITLE
Login

### PR DESCRIPTION
**Pull request by ghost**
Login components...

### CODE REVIEW

File: app/templates/app/board_view.html [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    {% for list in board_list %}
+    <li>
+        <div class="btn-group">
+            <!-- Display List -->
+            <button type="button" class="btn btn-light">{{list.title}}</button>
+            <button type="button" class="btn btn-light fa fa-ellipsis-h" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
+                <span class="sr-only">Toggle Dropdown</span>
+            </button>
+            <div class="dropdown-menu">
+                <a class="dropdown-item" href="../list/delete/{{ list.id }}">Delete</a>
```
Comment (mcandidier) [inline app/templates/app/board_view.html:18]: use url template_tag for this

File: app/templates/app/card_ajax.html [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+{% for card in cards %}
+<li>
+    <div class="btn-group">
+        <!-- Display List -->
+        <button type="button" class="btn btn-light">{{card.title}}</button>
+        <button type="button" class="btn btn-light fa fa-ellipsis-h" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
+            <span class="sr-only">Toggle Dropdown</span>
+        </button>
+        <div class="dropdown-menu">
+            <a class="dropdown-item" href="../list/cards/delete/{{card.id }}">Delete</a>
```
Comment (mcandidier) [inline app/templates/app/card_ajax.html:11]: use url template_tag for this

File: app/templates/app/card_edit.html [hunk 3]
```diff
@@ -0,0 +1,9 @@
+{% extends 'app/base.html' %}
+
+{% block content %}
+    <h2>New post</h2>
+    <form method="POST" class="">{% csrf_token %}
+        {{ form.as_p }}
```
Comment (mcandidier) [inline app/templates/app/card_edit.html:6]: render form field manually

File: app/templates/app/create_board.html [hunk 4]
```diff
@@ -0,0 +1,11 @@
+{% extends 'app/base.html' %}
+{% block content %}
+    <h2>Create Board</h2>
+    <form method="post" novalidate>
+        {% csrf_token %}
+        {% for forms in form %}
```
Comment (mcandidier) [inline app/templates/app/create_board.html:6]: doesn't make sense, make sure that your variable naming understandable, (loop forms in form?)

File: app/templates/app/login.html [hunk 5]
```diff
@@ -0,0 +1,14 @@
+{% extends 'app/base.html' %}
+{% block content %}
+    <h2>Login User</h2>
+    <form method="post" novalidate>
+        {% csrf_token %}
+        <div class="form-group">
```
Comment (mcandidier) [inline app/templates/app/login.html:6]: make sure to show form field errors as well.

File: app/views.py [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+        return render(self.request, self.template_name,{ 'form' : form})
+
+    def post(self, *args, **kwargs):
+        username = self.request.POST['username']
+        password = self.request.POST['password']
+        user = authenticate(self.request, username=username, password=password)
+        if user is not None:
+            login(self.request, user)
+            return HttpResponseRedirect('/')
+        #  else:
```
Comment (mcandidier) [inline app/views.py:31]: what happens if user is equal to None? make sure to handle errors

File: app/views.py [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+        form = self.form(request.POST)
+        # check whether it's valid:
+        if form.is_valid():
+            # if it is valid, save(commit=False) gets you a model object, then you can add your extra data
+            form = form.save(commit=False)
+            # add user data then save() it
+            form.user = request.user
+            form.save()
+            # redirect to a home url:
+            return HttpResponseRedirect('/')
```
Comment (mcandidier) [inline app/views.py:70]: do not use static path, practice to use url names

File: app/views.py [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+
+    template_name = 'app/board_view.html'
+    def get(self, *args, **kwargs):
+        board_id = kwargs.get('id')
+        # iterate board model, if data doesnt exist it will move you to 404 page
+        board = get_object_or_404(Board, id=board_id)
+        # iterate list model
+        board_list = BoardList.objects.filter(board=board_id)
+        # select list id
+        card_query = Card.objects.all()
```
Comment (mcandidier) [inline app/views.py:86]: this is wrong, filter cards by boardList

File: app/views.py [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+    Redirect the user to the selected board
+    """
+
+    template_name = 'app/board_view.html'
+    def get(self, *args, **kwargs):
+        board_id = kwargs.get('id')
+        # iterate board model, if data doesnt exist it will move you to 404 page
+        board = get_object_or_404(Board, id=board_id)
+        # iterate list model
+        board_list = BoardList.objects.filter(board=board_id)
```
Comment (mcandidier) [inline app/views.py:84]: wrong, make sure to filter boardList object by list id
you get board object using board_id, then get boardlist object by board_id?

File: app/templates/app/card_ajax.html [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+            <a class="dropdown-item" href="../list/card/edit/{{card.id }}">{{card.id }} edit</a>
+    </div>
+</li>
+{% empty %}
+<p>no cards</p>
+{% endfor %}
+
```
Comment (mcandidier) [inline app/templates/app/card_ajax.html:18]: remove empty lines

File: app/templates/app/create_card.html [hunk 11]
```diff
@@ -0,0 +1,11 @@
+{% extends 'app/base.html' %}
+{% block content %}
+    <h2>Create Card</h2>
+    <form method="post" novalidate>
+        {% csrf_token %}
+        {% for forms in form %}
```
Comment (mcandidier) [inline app/templates/app/create_card.html:6]: looping forms inside form? wrong grammar

File: app/templates/app/create_list.html [hunk 12]
```diff
@@ -0,0 +1,11 @@
+{% extends 'app/base.html' %}
+{% block content %}
+    <h2>Create List</h2>
+    <form method="post" novalidate>
+        {% csrf_token %}
+        {% for list in list_view %}
```
Comment (mcandidier) [inline app/templates/app/create_list.html:6]: make sure to named it correctly, doesn't make sense list in list_view

File: app/views.py [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+        return render(self.request, self.template_name, {'board': board, 'board_list' : board_list, 'card_views' : card_query})
+
+
+class AddList(TemplateView):
+    """
+    Let the user add new list
+    """
+
+
+    list_view = ListForm
```
Comment (mcandidier) [inline app/views.py:96]: named it to forms rather than list_view

Comment (mcandidier) [reply app/views.py:96]: make sure to name it based on our standards

File: app/views.py [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+    template_name = 'app/create_list.html'
+    def get(self, *args, **kwargs):
+        list_view = self.list_view()
+        return render(self.request, self.template_name, {'list_view': list_view})
+
+    def post(self, *args, **kwargs):
+        board_id = kwargs.get('id')
+        board = Board.objects.get(id=board_id)
+        # create a form instance and populate it with data from the request:
+        list_view = self.list_view(self.request.POST)
```
Comment (mcandidier) [inline app/views.py:106]: named it correctly, 'form'

File: app/views.py [hunk 15]
```diff
@@ -0,0 +1,231 @@
...
+            # redirect to a board views url:
+            return redirect('board_view', board_id)
+        return render(self.request, self.template_name, {'list_view': list_view})
+
+
+class EditList(TemplateView):
+    """
+    Let the user edit list
+    """
+
```
Comment (mcandidier) [inline app/views.py:121]: add single space after docstring

File: app/views.py [hunk 16]
```diff
@@ -0,0 +1,231 @@
...
+    template_name = 'app/list_edit.html'
+    def get(self, *args, **kwargs):
+        list_id = kwargs.get('id')
+        _list = get_object_or_404(BoardList, id=list_id)
+        form = self.form(instance = _list)
+        return render(self.request, self.template_name, {'form' : form})
+
+    def post(self, *args, **kwargs):
+        list_id = kwargs.get('id')
+        _list = get_object_or_404(BoardList, id=list_id)
```
Comment (mcandidier) [inline app/views.py:133]: don't named variable by underscore

File: app/views.py [hunk 17]
```diff
@@ -0,0 +1,231 @@
...
+        card_id = kwargs.get('id')
+        card = get_object_or_404(Card, id=card_id)
+        form = self.form(instance = card)
+        return render(self.request, self.template_name, {'form' : form})
+
+    def post(self, *args, **kwargs):
+        card_id = kwargs.get('id')
+        card = get_object_or_404(Card, id=card_id)
+        form = self.form(self.request.POST, instance = card)
+        if form.is_valid():
```
Comment (mcandidier) [inline app/views.py:186]: make sure to handle form errors

File: app/views.py [hunk 18]
```diff
@@ -0,0 +1,231 @@
...
+    Let the user delete list
+    """
+
+
+    template_name = 'board_view'
+    def get(self, request, *args, **kwargs):
+        list_id = kwargs.get('id')
+        _list = get_object_or_404(BoardList, id=list_id)
+        _list.delete()
+        return redirect(self.template_name, _list.board.id)
```
Comment (mcandidier) [inline app/views.py:202]: wrong implementation, redirect using template?

File: app/views.py [hunk 19]
```diff
@@ -0,0 +1,231 @@
...
+class DeleteList(TemplateView):
+    """
+    Let the user delete list
+    """
+
+
+    template_name = 'board_view'
+    def get(self, request, *args, **kwargs):
+        list_id = kwargs.get('id')
+        _list = get_object_or_404(BoardList, id=list_id)
```
Comment (mcandidier) [inline app/views.py:200]: remove underscore, bad habit

**Comment by mcandidier** [comment]
remove .pyc files
remove environment files

File: app/templates/app/board_view.html (hunk 1 above)
Comment (ghost) [reply app/templates/app/board_view.html:18]: Done sir

File: app/templates/app/card_ajax.html (hunk 2 above)
Comment (ghost) [reply app/templates/app/card_ajax.html:11]: Done sir

File: app/templates/app/card_edit.html (hunk 3 above)
Comment (ghost) [reply app/templates/app/card_edit.html:6]: Done sir

File: app/templates/app/create_board.html (hunk 4 above)
Comment (ghost) [reply app/templates/app/create_board.html:6]: Done sir

File: app/templates/app/login.html (hunk 5 above)
Comment (ghost) [reply app/templates/app/login.html:6]: Done sir

File: app/views.py (hunk 6 above)
Comment (ghost) [reply app/views.py:31]: Done sir

File: app/views.py (hunk 7 above)
Comment (ghost) [reply app/views.py:70]: Done sir

File: app/views.py (hunk 8 above)
Comment (ghost) [reply app/views.py:86]: Done sir

File: app/views.py (hunk 9 above)
Comment (ghost) [reply app/views.py:84]: Done sir

File: app/templates/app/card_ajax.html (hunk 10 above)
Comment (ghost) [reply app/templates/app/card_ajax.html:18]: Done sir

File: app/templates/app/create_card.html (hunk 11 above)
Comment (ghost) [reply app/templates/app/create_card.html:6]: Done sir

File: app/templates/app/create_list.html (hunk 12 above)
Comment (ghost) [reply app/templates/app/create_list.html:6]: Done sir

File: app/views.py (hunk 13 above)
Comment (ghost) [reply app/views.py:96]: Done sir

File: app/views.py (hunk 14 above)
Comment (ghost) [reply app/views.py:106]: Done sir

**Pull request closed without merging**